lib/fluent/plugin/filter_grep.rb

```diff
@@ -54,65 +59,139 @@ def initialize
       config_param :key, :string
       desc "The regular expression."
       config_param :pattern do |value|
-        if value.start_with?("/") and value.end_with("/")
+        if value.start_with?("/") and value.end_with?("/")
           Regexp.compile(value[1..-2])
         else
           Regexp.compile(value)
         end
       end
     end
 
-    # for test
-    attr_reader :_regexps
-    attr_reader :_excludes
+    config_section :and, param_name: :and_conditions, multi: true do
+      config_section :regexp, param_name: :regexps, multi: true do
+        desc "The field name to which the regular expression is applied."
+        config_param :key, :string
+        desc "The regular expression."
+        config_param :pattern do |value|
+          if value.start_with?("/") and value.end_with?("/")
+            Regexp.compile(value[1..-2])
+          else
+            Regexp.compile(value)
+          end
+        end
+      end
+      config_section :exclude, param_name: :excludes, multi: true do
+        desc "The field name to which the regular expression is applied."
+        config_param :key, :string
+        desc "The regular expression."
+        config_param :pattern do |value|
+          if value.start_with?("/") and value.end_with?("/")
+            Regexp.compile(value[1..-2])
+          else
+            Regexp.compile(value)
+          end
+        end
+      end
+    end
+
+    config_section :or, param_name: :or_conditions, multi: true do
+      config_section :regexp, param_name: :regexps, multi: true do
+        desc "The field name to which the regular expression is applied."
+        config_param :key, :string
+        desc "The regular expression."
+        config_param :pattern do |value|
+          if value.start_with?("/") and value.end_with?("/")
+            Regexp.compile(value[1..-2])
+          else
+            Regexp.compile(value)
+          end
+        end
+      end
+      config_section :exclude, param_name: :excludes, multi: true do
+        desc "The field name to which the regular expression is applied."
+        config_param :key, :string
+        desc "The regular expression."
+        config_param :pattern do |value|
+          if value.start_with?("/") and value.end_with?("/")
+            Regexp.compile(value[1..-2])
+          else
+            Regexp.compile(value)
+          end
+        end
+      end
+    end
 
     def configure(conf)
       super
 
-      rs = {}
       (1..REGEXP_MAX_NUM).each do |i|
         next unless conf["regexp#{i}"]
         key, regexp = conf["regexp#{i}"].split(/ /, 2)
         raise Fluent::ConfigError, "regexp#{i} does not contain 2 parameters" unless regexp
-        raise Fluent::ConfigError, "regexp#{i} contains a duplicated key, #{key}" if rs[key]
-        rs[key] = Regexp.compile(regexp)
+        raise Fluent::ConfigError, "regexp#{i} contains a duplicated key, #{key}" if @_regexp_and_conditions[key]
```

Comment:
How about logging info or warning message for top level setting?
`Top level multiple <regexp> is intepreted as 'and' condition`.